ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java

```diff
@@ -77,87 +72,100 @@ public String getName() {
-      if (!validationResult.isValid()) {
-        return new JsonRpcErrorResponse(
-            id, convertTransactionInvalidReason(validationResult.getInvalidReason()));
+
+      if (!privateTransaction.getPrivateFrom().equals(Bytes.fromBase64String(enclavePublicKey))) {
+        return new JsonRpcErrorResponse(id, PRIVATE_FROM_DOES_NOT_MATCH_ENCLAVE_PUBLIC_KEY);
       }
 
-      final String enclaveKey =
-          privacyController.sendTransaction(privateTransaction, enclavePublicKey);
-      final Transaction privacyMarkerTransaction;
-      if (privateTransaction.getPrivacyGroupId().isPresent()) {
-        PrivacyGroup privacyGroup = null;
-        try {
-          privacyGroup =
-              privacyController.retrievePrivacyGroup(
-                  privateTransaction.getPrivacyGroupId().get().toBase64String(),
-                  enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser()));
-        } catch (final EnclaveClientException e) {
-          // it is an onchain group
+      PrivacyGroup privacyGroup = null;
+      final Optional<Bytes> maybePrivacyGroupId = privateTransaction.getPrivacyGroupId();
+      if (onchainPrivacyGroupsEnabled) {
+        if (!maybePrivacyGroupId.isPresent()) {
+          // on-chain private transaction does not contain privacy group ID
+          return new JsonRpcErrorResponse(
+              id, JsonRpcError.PRIVACY_GROUP_ID_NOT_AVAILABLE_WHILE_ON_CHAIN_PRIVACY_IS_ENABLED);
         }
+        privacyGroup =
+            privacyController.retrieveOnChainPrivacyGroup(
+                maybePrivacyGroupId.get(), enclavePublicKey);
         if (privacyGroup == null
-            || !privacyGroup
-                .getMembers()
-                .contains(enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser()))) {
-
-          if (!isOnchainPrivacyGroupEnabled) {
-            return new JsonRpcErrorResponse(id, ONCHAIN_PRIVACY_GROUP_NOT_ENABLED);
-          }
-
-          final Optional<String> addPayloadEnclaveKey =
-              privacyController.buildAndSendAddPayload(
-                  privateTransaction,
-                  enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser()));
-
-          privacyMarkerTransaction =
-              privacyController.createPrivacyMarkerTransaction(
-                  buildCompoundKey(enclaveKey, addPayloadEnclaveKey),
-                  privateTransaction,
-                  Address.ONCHAIN_PRIVACY);
-        } else {
-          if (isOnchainPrivacyGroupEnabled) {
-            return new JsonRpcErrorResponse(id, OFFCHAIN_PRIVACY_GROUP_NOT_ENABLED);
-          }
-
-          privacyMarkerTransaction =
-              privacyController.createPrivacyMarkerTransaction(enclaveKey, privateTransaction);
+            && !privacyController.isGroupAdditionTransaction(privateTransaction)) {
+          // on-chain privacy enabled, but on-chain privacy group does not exist and is not an
+          // on-chain privacy group creation
```

Comment:
I think we can remove this comment. The code is clear on what is happening.

Reply:
Done